gtest/src/accounts.rs

```diff
@@ -0,0 +1,169 @@
+    }
+}
+
+thread_local! {
+    static ACCOUNT_STORAGE: RefCell<HashMap<ProgramId, Balance>> = RefCell::new({
+        let mut storage = HashMap::new();
+        init_default_accounts(&mut storage);
+        storage
+    });
+}
+
+#[derive(Debug)]
+struct Balance {
+    amount: Value,
+}
+
+impl Balance {
+    fn new(amount: Value) -> Self {
+        Self { amount }
+    }
+
+    fn balance(&self) -> Value {
+        self.amount
+    }
+
+    fn reducible_balance(&self) -> Value {
+        self.amount - EXISTENTIAL_DEPOSIT
+    }
+
+    fn decrease(&mut self, amount: Value) {
+        self.amount -= amount;
+    }
+
+    fn increase(&mut self, amount: Value) {
+        self.amount += amount;
+    }
+}
+
+pub(crate) struct Accounts;
+
+impl Accounts {
+    // Checks if account by program id exists.
+    pub(crate) fn is_exist(id: ProgramId) -> bool {
+        Self::balance(id) != 0
+    }
+
+    // Returns account balance.
+    pub(crate) fn balance(id: ProgramId) -> Value {
+        ACCOUNT_STORAGE.with_borrow(|storage| {
+            storage
+                .get(&id)
+                .map(|balance| balance.balance())
+                .unwrap_or_default()
+        })
+    }
+
+    // Returns account reducible balance.
+    pub(crate) fn reducible_balance(id: ProgramId) -> Value {
+        ACCOUNT_STORAGE.with_borrow(|storage| {
+            storage
+                .get(&id)
+                .map(|balance| balance.reducible_balance())
+                .unwrap_or_default()
+        })
+    }
+
+    // Decreases account balance.
+    pub(crate) fn decrease(id: ProgramId, amount: Value, keep_alive: bool) {
+        ACCOUNT_STORAGE.with_borrow_mut(|storage| {
+            if let Some(balance) = storage.get_mut(&id) {
+                if keep_alive && balance.reducible_balance() < amount {
+                    panic!(
+                        "Not enough balance to decrease, reducible: {}, value: {amount}",
+                        balance.reducible_balance(),
+                    );
+                }
+                if !keep_alive && balance.balance() < amount {
+                    panic!(
+                        "Not enough balance to decrease, total: {}, value: {amount}",
+                        balance.balance(),
+                    );
+                }
+
+                balance.decrease(amount);
+                if balance.balance() < EXISTENTIAL_DEPOSIT {
+                    log::debug!(
+                        "Removing account {id:?} with balance {} below the existential deposit",
+                        balance.balance()
+                    );
+                    storage.remove(&id);
+                }
+            }
+        });
+    }
+
+    // Increases account balance.
+    pub(crate) fn increase(id: ProgramId, amount: Value) {
+        ACCOUNT_STORAGE.with_borrow_mut(|storage| {
+            let balance = storage.entry(id).or_insert(Balance::new(0));
```

Comment:
thinking of future refactoring with introducing a separate error type I'd not call `or_insert` unless the check below that account has value gt ED is passed.
```rust
            // first check
            if Self::balance(id) + amount < EXISTENTIAL_DEPOSIT { // panic here }
            // and then this increase
            let balance = storage.entry(id).or_insert(Balance::new(0));
```

Reply:
Fixed